### PR TITLE
Clang tidy

### PR DESCRIPTION
**Pull request by sithhell**
Fixes as reported by clang-tidy after enabling the following checks:

    modernize-use-nullptr
    misc-use-after-move
    misc-virtual-near-miss
    misc-multiple-statement-macro
    misc-move-constructor-init
    misc-move-forwarding-reference
    misc-assert-side-effect
    misc-dangling-handle
    misc-move-constructor-init
    misc-move-forwarding-reference
    misc-non-copyable-objects
    misc-forwarding-reference-overload

### CODE REVIEW

File: hpx/parallel/algorithms/for_loop.hpp [hunk 1]
```diff
@@ -206,19 +206,19 @@ namespace hpx { namespace parallel { inline namespace v2
...
 
+                args_type args =
+                    hpx::util::forward_as_tuple(std::forward<Ts>(ts)...);
+
                 return util::partitioner<ExPolicy>::call_with_index(
                     policy, first, size, stride,
                     part_iterations<F, S, args_type>{
-                        std::forward<F>(f), stride, std::move(args)
+                        std::forward<F>(f), stride, args
                     },
```
Comment (hkaiser) [inline hpx/parallel/algorithms/for_loop.hpp]: AFAICS, this code causes the data to be copied twice. Once when creating `args` and once while calling `call_with_index`. Is there any way to avoid that?

File: src/runtime/agas/server/route.cpp [hunk 2]
```diff
@@ -75,6 +75,7 @@ namespace hpx { namespace agas { namespace server
             addr.address_ = g.lva();
         }
 
+        naming::id_type source = p.source_id();
```
Comment (hkaiser) [inline src/runtime/agas/server/route.cpp]: What is the rationale of this change?

File: src/runtime/resource/detail/detail_partitioner.cpp [hunk 3]
```diff
@@ -505,7 +505,6 @@ namespace hpx { namespace resource { namespace detail
         affinity_data_.set_num_threads(new_pu_nums.size());
         affinity_data_.set_pu_nums(std::move(new_pu_nums));
         affinity_data_.set_affinity_masks(std::move(new_affinity_masks));
-        affinity_data_.init_cached_pu_nums(new_pu_nums.size());
```
Comment (hkaiser) [inline src/runtime/resource/detail/detail_partitioner.cpp]: What is the rationale of this change? is that related to the clang-tidy checks?

File: src/runtime/threads/threadmanager.cpp [hunk 4]
```diff
@@ -286,7 +286,7 @@ namespace hpx { namespace threads
             "default",
             "low",
             "normal",
-            "high (recursive)"
+            "high (recursive)",
```
Comment (hkaiser) [inline src/runtime/threads/threadmanager.cpp]: Excellent catch!

File: tools/clang-tidy.sh [hunk 5]
```diff
@@ -38,15 +38,62 @@ NUM_FILES=$(echo "${FILES}" | wc -l)
...
+CHECKS="$CHECKS,modernize-use-nullptr"
+CHECKS="$CHECKS,misc-use-after-move"
+CHECKS="$CHECKS,misc-virtual-near-miss"
+CHECKS="$CHECKS,misc-multiple-statement-macro"
+CHECKS="$CHECKS,misc-move-constructor-init"
+CHECKS="$CHECKS,misc-move-forwarding-reference"
+CHECKS="$CHECKS,misc-assert-side-effect"
+CHECKS="$CHECKS,misc-dangling-handle"
+CHECKS="$CHECKS,misc-move-constructor-init"
+CHECKS="$CHECKS,misc-move-forwarding-reference"
```
Comment (hkaiser) [inline tools/clang-tidy.sh]: These two are duplicated.

File: hpx/parallel/algorithms/for_loop.hpp (hunk 1 above)
Comment (sithhell) [reply hpx/parallel/algorithms/for_loop.hpp]: Yes, that's a potential problem indeed. I don't know how to avoid that. `args` is needed twice, once for `part_iterations` and then in the finalizer.

File: src/runtime/agas/server/route.cpp (hunk 2 above)
Comment (sithhell) [reply src/runtime/agas/server/route.cpp]: `p` is moved here: https://github.com/STEllAR-GROUP/hpx/pull/2982/files/69463c83e47825205fbb10147b2d97b6726e1242#diff-fe4bb9d1c3c7a064a0b4dd4bdae63198L87

That means that the `source_id` used later on to update the cache entry is being used from a moved from object.

File: src/runtime/resource/detail/detail_partitioner.cpp (hunk 3 above)
Comment (sithhell) [reply src/runtime/resource/detail/detail_partitioner.cpp]: Yes, `new_pu_nums` was moved a few lines above. When looking at the code of `affinity_data`, `init_cached_pu_nums` does nothing in this specific case anyway: https://github.com/STEllAR-GROUP/hpx/blob/master/src/runtime/threads/policies/affinity_data.cpp#L324

File: src/runtime/agas/server/route.cpp (hunk 2 above)
Comment (hkaiser) [reply src/runtime/agas/server/route.cpp]: Right, thanks for catching this.

File: hpx/parallel/algorithms/for_loop.hpp (hunk 1 above)
Comment (hkaiser) [reply hpx/parallel/algorithms/for_loop.hpp]: Couldn't `args_type` be a tuple of references instead?

**Comment by sithhell** [comment]
*@hkaiser* commented on this pull request.
------------------------------

In hpx/parallel/algorithms/for_loop.hpp
<https://github.com/STEllAR-GROUP/hpx/pull/2982#discussion_r148835398>:

>                  return util::partitioner<ExPolicy>::call_with_index(
                     policy, first, size, stride,
                     part_iterations<F, S, args_type>{
-                        std::forward<F>(f), stride, std::move(args)
+                        std::forward<F>(f), stride, args
                     },

Couldn't args_type be a tuple of references instead?


No we'd get dangling references then. We need to decay copy the arguments
to stay in scope long enough.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/STEllAR-GROUP/hpx/pull/2982#discussion_r148835398>, or mute
the thread

File: hpx/parallel/algorithms/for_loop.hpp (hunk 1 above)
Comment (sithhell) [reply hpx/parallel/algorithms/for_loop.hpp]: No, they need to decay-copied at least once to stay in scope long enough.

Comment (hkaiser) [reply hpx/parallel/algorithms/for_loop.hpp]: Ok, so let's copy them once and not thrice.

Comment (sithhell) [reply hpx/parallel/algorithms/for_loop.hpp]: I wouldn't know how...
The first one, is forwarding the arguments into a local tuple, which then needs to be passed to the iteration function and the exit function, which both need a copy of the arguments.

Comment (hkaiser) [reply hpx/parallel/algorithms/for_loop.hpp]: Well, the exit function is guaranteed to outlive everything else, right? So why not hold the copy there and let the iteration refer to those copies? Would that work?

Comment (sithhell) [reply hpx/parallel/algorithms/for_loop.hpp]: Hmm, this would require some significant refactoring of the partitioner. I would prefer doing that in another PR though, but I guess it would work in principle

**Comment by biddisco** [comment]
This PR looks like it should be merged. I will create an Issue to remember the partitioner copy problem and then someone can/should merge this.

File: hpx/parallel/algorithms/for_loop.hpp (hunk 1 above)
Comment (sithhell) [reply hpx/parallel/algorithms/for_loop.hpp]: Ok, the issue of having multiple copies should be fixed with the latest commit.

**Comment by sithhell** [comment]
All comments have been addressed. This should be good to go now.

File: hpx/parallel/algorithms/for_loop.hpp (hunk 1 above)
Comment (hkaiser) [reply hpx/parallel/algorithms/for_loop.hpp]: Perfect, thanks!
